### ISSUE TITLE
Multiarch Docker container

### ISSUE THREAD

**Issue by githubsaturn** [certbot/certbot#8225]
Currently the docker containers are built for various architectures, but they are tagged separately. This means that an end user needs to know their architecture tag, when they want to run `docker run certbot/certbot:architecture-tag`

A more appropriate way to do this is by using Docker manifest that packages multiple images under the same tag. Most widely used images are currently using this technique. See nginx for example:
![image](https://user-images.githubusercontent.com/10582082/90989547-8a6d4d80-e568-11ea-9e41-1c683a0f49c2.png)

---

Moving to this new behavior is quite simple using buildx. A simple build and push that used to be like this:
```
docker build -t myuser/myimage:tag .
docker push myuser/myimage:tag
```
will change to 
```
export DOCKER_CLI_EXPERIMENTAL=enabled
docker buildx create --name mybuilder
docker buildx use mybuilder
docker buildx build --platform linux/amd64,linux/arm64,linux/arm/v7 -t myuser/myimage:tag --push .
``` 

It is worth noting that the buildx is quite stable. It is, in fact, powering the standard `docker build` command. The only reason that is part of the experimental package, AFAIK, is that Docker is not committing to keeping the same CLI signature - this should be fine for a build process. They haven't change the CLI args in the past year, even if they do, Cerbot docker needs to update the build script. No risk to the builds.

**Comment by ohemorange** [comment]
@bmw I think I saw a discussion going by about our docker builds but I wasn't following it so I'll leave this in your capable hands

**Comment by bmw** [comment]
While I'm interested in doing both providing a manifest and using `buildx`, as you said, despite them being quite stable, Docker still flags them as experimental features so I think we probably won't adopt them until they remove that designation.

**Comment by githubsaturn** [comment]
Generally it's okay to use experimental features in the release pipeline. If things break due to a version upgrade of Docker, you can always go and change the flag to match the new requirement and be done with it.

IMHO the benefit and value it provides vastly exceed the potential headache of a small modification in the build script.

**Most, if not all, other "high profile" Docker images have proper multiarch images**. As an alternative, you can use `docker manifest` command. It is not as seamless as `buildx` - but it'll get the job done. So here are the changes that you need to make:

1) Build and push all images exactly the way you currently do.

2) Add this script as the last step of your build process after all images are pushed to DockerHub:
```
docker manifest create $CERTBOT_IMAGE_NAME:$CERTBOT_VERSION \
                       $CERTBOT_IMAGE_NAME:amd64-$CERTBOT_VERSION \
                       $CERTBOT_IMAGE_NAME:arm64v8-$CERTBOT_VERSION \
                       $CERTBOT_IMAGE_NAME:arm32v6-$CERTBOT_VERSION


docker manifest push $CERTBOT_IMAGE_NAME:$CERTBOT_VERSION
```

**Comment by bmw** [comment]
Thanks for the follow up info. I agree there are a lot of high profile projects using these experimental features, but I still have the same hesitation for now. Maybe we'll pick this up in the future even if it's still flagged as experimental when we have more spare time.

I'll also mention that `docker manifest` is also an experimental feature that "should not be used in production environments". See https://docs.docker.com/engine/reference/commandline/manifest/.

**Comment by danielwagn3r** [comment]
From a usability point of view it would be desirable to have the same tags used for all supported platforms.

**Comment by ldrrp** [comment]
I came here looking to report this issue with nginx as an example. So happy to see this ticket exactly how i would have posted it. +1 this. I have a project i would like to support amd64 and raspberry pi's but i would really hate to have to make different docker-compose files or a fork just for arm32 and arm64. Would be nice for this to work like nginx container

**Comment by MC-DeltaT** [comment]
Is Docker Buildx even experimental any more? If you look at the [summary page](https://docs.docker.com/buildx/working-with-buildx/) it doesn't mention anything about being experimental or to avoid using it in production or whatever. I don't think Docker is discouraging us from using it.

The thing is that multi-arch support is such an amazing feature that, in my opinion, it's a no-brainer to use it. And note that this is the opinion of _many_ others too - if you look at basically any popular Docker image, you'll see it uses multi-arch builds. If Docker pull support for this feature because it's "experimental", then don't worry, the majority of Docker Hub is going down with you.

The problem with not having multi-arch builds is that it cripples everyone who uses your images in an application they want to be cross-platform.
For example, I want to portably run certbot in my app on Raspberry Pi, while being able to develop on an x86 Windows machine. Well I pretty much just can't, because there's no multi-arch support. And then it's a pain to use certbot any other way, because basically the other only supported method is snap... Which doesn't work in Docker (as far as I'm aware)... And isn't supported on Windows...
On the other hand, with multi-arch builds, everything works nicely by default, no messing around required.

**Comment by Roming22** [comment]
One more vote to enable multi-arch. Not just because it's more convenient for the user, but also because it simplifies the generation and reuse of deployments.yml for k8s.

**Comment by humanoid2050** [comment]
Buildx is stable. It now ships as part of Docker Desktop and Docker Engine when installed via most package managers. https://docs.docker.com/build/install-buildx/
Adding multiarch support is a lost cost/high return improvement, especially as Arm-based servers become more prevalent.

**Comment by bmw** [comment]
Nice! It does seem like buildx is no longer flagged as experimental. [`docker manifest` still is flagged that way](https://docs.docker.com/engine/reference/commandline/manifest/) unfortunately though with [blog posts from Docker in the last 7 months](https://www.docker.com/blog/how-to-rapidly-build-multi-architecture-images-with-buildx/) containing text like:

> Note: However, you should only use the docker manifest command in testing — not production. This command is experimental. We’re continually tweaking functionality and any associated UX while making docker manifest production ready.

**Comment by humanoid2050** [comment]
Using buildx with multiarch support actually means you don't have to invoke `docker manifest`. So long as you have a registry to use as a destination with the `--push` flag, then buildx does everything for you in one shot.

**Comment by bmw** [comment]
That's even nicer. Thanks for all the info.

If that's the case, I'd be open to making this change. If someone wants to try and write a PR for it, our Docker tooling is under `tools/docker` and is used by our Azure Pipelines configuration which is under `.azure-pipelines`. If the workflow of the `tools/docker` scripts or Docker pipelines would have to change much, I'd encourage any potential PR author to ask about those changes here early and get a +1 before spending a lot of time on the implementation details.

**Comment by humanoid2050** [comment]
Might be a fairly light lift? Couple questions to make sure I understand the design patterns of this project.
1. I see that `core/Dockerfile` and `plugin/Dockerfile` both copy QEMU static binaries into the build images. What is the purpose of this? Are there binaries inside the images that do not match the nominal architecture? 
2. What is the right way to tell the azure to just do steps without a matrix strategy? I'm looking at `packaging-jobs.yml` and the one-shot nature of buildx probably means there is no need to have separate jobs for each architecture.
3. How "isolated" is the azure pipeline? I see that all the docker images are pulled from the local image cache and stored in an `images.tar` file. The one downside to buildx automation is that manifests and multiarch images are incompatible with the local image cache. As such, the azure pipeline will probably need a temporary docker registry to act as a destination for the build jobs and a pull source during the run and testing jobs. 

I've never used azure pipelines so I'm a little out of my depth on that front.

**Comment by bmw** [comment]
Sorry for the couple day delay. It's been a busy week.

1. Yes. Azure always runs for us on x86-64 and building the ARM Docker images requires [building](https://dev.azure.com/certbot/certbot/_build/results?buildId=6385&view=logs&j=fdd3565a-f3c6-5154-eca9-9ae03666f7bd&t=5dbd9851-46a4-524f-73a8-4028241afcde&l=343) or [selecting](https://dev.azure.com/certbot/certbot/_build/results?buildId=6385&view=logs&j=40dc3f3b-10cb-5f7f-1c5a-019f33f252fd&t=95f1458e-ebc6-55bb-9757-5c263b6e09d4&l=281) code for the right architecture.
2. If you just omit the `strategy:` and `matrix:` sections, I think it should work. For instance, see [(Windows) `installer_build` job](https://github.com/certbot/certbot/blob/057524aa52fce8a300e48fcdaa04ea41acac9e09/.azure-pipelines/templates/jobs/packaging-jobs.yml#L54) that's under the [`docker_build` job](https://github.com/certbot/certbot/blob/057524aa52fce8a300e48fcdaa04ea41acac9e09/.azure-pipelines/templates/jobs/packaging-jobs.yml#L2).
3. Each "job" in a pipleline is a fresh environment. I think this might be the hardest part about doing this. I'm not sure the best place for us to temporarily store these images. Azure has a [container registry](https://azure.microsoft.com/en-us/products/container-registry), but that seems to be a paid service, even for open source projects. While not ideal, another potential option if `buildx` allows it is for us to build, test, and then push the Docker images all within a single job.

**Comment by humanoid2050** [comment]
I think I'm starting to wrap my head around this. What exactly does the `docker_run` job do? I see a `docker run` command for all the collected image tags. I assume if the container returns a non-zero exit code the whole pipeline halts and that indicates a failure of some kind? Also, it looks like the `docker_run` job only loads the amd64 `images.tar` artifact. Is that correct? 

Stepping back for just a moment from the `docker buildx build --push` pattern, I also see that there is a `docker buildx imagetools create` method which seems to be suitable for stitching multiple images in a registry into a single manifest. That might be a more direct path to building a multiarch image than restructuring the azure pipeline. All existing images could be pushed with their fully self-describing tags, and one new step would be added to create the manifest with the appropriate tag at the very end. I will dig into it a bit.

**Comment by humanoid2050** [comment]
Looks like `docker buildx imagetools create` is the way to go. I think changes will be isolated to `deploy-stage.yml` and `deploy.sh`. The net result would be separating the arch-specific image uploads from the creation of the base tags representing multiarch manifest "images".

**Comment by bmw** [comment]
> I think I'm starting to wrap my head around this. What exactly does the `docker_run` job do? I see a `docker run` command for all the collected image tags. I assume if the container returns a non-zero exit code the whole pipeline halts and that indicates a failure of some kind? Also, it looks like the `docker_run` job only loads the amd64 `images.tar` artifact. Is that correct? 

Yeah I think you've got it. `docker_run` is meant to just be a smoke test of the previously built Docker images. It runs the simple command `certbot plugins --prepare` which just has Certbot try to find, initialize, and prepare its plugins like they're about to be used and then exit. If `docker_run` fails, the pipeline is halted and doesn't continue to the later stages which publish the images to Docker Hub. `docker_run` does just run for amd64 because it's difficult for us to test the other architectures with our Azure Pipelines setup and only x86-64 machines available.

`docker buildx imagetools create` sounds promising :eyes: I didn't dig into it too deeply, but I don't immediately see any problems with that approach. Thanks so much for looking into this and please let me know if you have any other questions.

**Comment by humanoid2050** [comment]
Are there cases where the full suite of architectures are not built? The fact that `build.sh`, `deploy.sh`, and the azure-pipelines are all separately coded with references to the various architectures feels fragile. 

I have a draft commit at https://github.com/humanoid2050/certbot. It worked once and then got weird. I seem to be unable to run the build script for non-amd64 architectures. The plugin stage complains that the base certbot image can't be found, even though it is in my local cache. Does that error seem familiar for you?

**Comment by bmw** [comment]
> Are there cases where the full suite of architectures are not built? The fact that `build.sh`, `deploy.sh`, and the azure-pipelines are all separately coded with references to the various architectures feels fragile. 

In CI there used to be cases of that, but now they're always built. I see what you mean, but I also think being able to build a subset of architectures when playing with things manually is kind of nice 🤷 

> I have a draft commit at https://github.com/humanoid2050/certbot. It worked once and then got weird. I seem to be unable to run the build script for non-amd64 architectures. The plugin stage complains that the base certbot image can't be found, even though it is in my local cache. Does that error seem familiar for you?

That error does not seem familiar to me unfortunately.

**Comment by humanoid2050** [comment]
@bmw Not sure what to do about it yet, but the build weirdness seems to be a function of docker 32.0 changing to using buildkit by default for docker build. https://docs.docker.com/engine/release-notes/23.0/
sooo.... heads up?

**Comment by humanoid2050** [comment]
Well I figured it out. Calls to `docker build` for the plugin need this monstrosity added to the call

`--build-context "${DOCKER_HUB_ORG}/certbot:${TARGET_ARCH}-${TAG_BASE}"=docker-image://"${DOCKER_HUB_ORG}/certbot:${TARGET_ARCH}-${TAG_BASE}" \`

I guess the question is whether you want to commit to docker > 23.0 or lock your version of docker in azuer pipelines somehow.

**Comment by bmw** [comment]
Ooo thanks for letting me know. Neither my docker desktop installation or my test server's docker install from Ubuntu's snap store has updated yet.

I definitely don't want to pin back docker. I think for now we should either set the environment variable `DOCKER_BUILDKIT=0` in our scripts or switch the commands to `docker buildx build`. My understanding based on their changelog is that'd make the scripts behave similarly across their change in 23.0.

I think you should feel free to take the `DOCKER_BUILDKIT=0` approach for now if you want. If you think we should switch to `docker buildx build` (as I think I do as it's Docker's new default) and you want to do that work, I'd love that, but if it gets gnarly I think I'd probably want to see that work in a separate PR from the one adding a manifest to make things easier to review.

**Comment by humanoid2050** [comment]
PR #9586 is ready. It (temporarily) forces legacy docker builder behavior using `DOCKER_BUILDKIT=0`. It also adds a new stage to the azuer pipelines which calls a dedicated script for creating the multiarch images. I'm not really pleased with the separation of the base images from the multiarch manifest creation, but I don't see a better way forward without a more comprehensive conversion to buildkit and/or a deeper restructuring of the pipeline to remove the multiarch strategy elements.

**Comment by humanoid2050** [comment]
I've been considering next steps with BuildKit integration and looking at the comments from the PR. Do you want to keep the architecture-specific tags long term, or start deferring to the `docker pull --platform <architecture>` method for getting images of a specific (non-native) architecture? 

In principle, letting go of architecture-specific tags would make it possible to fire a single `docker build --platform <arch1>,<arch2>` invocation that bypasses per-architecture build and push operations. Separating build and push with a test stage does become more difficult, but in principle it would also be possible to bake a test step into the Dockerfile that produces no artifacts but fails the build if not successful.  I'm not sure where this project falls on the spectrum of backwards compatibility vs. progress.

**Comment by bmw** [comment]
We can potentially break backwards compatibility, however, it'd obviously be ideal not to. The other reason I'm hesitant to combine all the architectures into one build on top of that and increased difficulty testing is that the [Docker arm32v6 build is the bottleneck of the pipeline](https://dev.azure.com/certbot/certbot/_build/results?buildId=6444&view=logs&s=1dfbc15b-7d0f-52a9-b1da-b17592bf94f8) right now so I'd rather not do more work in that same build.

If we did want to go ahead and rely on `docker pull --platform` though, my next question would be how long ago did Docker add that flag and has it made it to the versions of Docker entombed in LTS Linux distros?

**Comment by humanoid2050** [comment]
It's always possible to build just a specific subset of architectures. In terms of real compute time, is the matrix strategy currently in use in the azure pipeline a parallel approach? As a corollary, how much parallelization does azure pipelines allow in practice? On my desktop with plenty of cores, use of buildkit means that all architectures get launched simultaneously, effectively resulting in a build time equivalent only to the slowest build.

I'm honestly not sure when the platform flag was added to `docker pull`. A quick search didn't turn up any clues either. I'm actually not against having per-arch tags, I just wanted to ask the question to see if there were opportunities to simplify.

I'm tinkering with some different approaches to get the most out of buildkit, and will probably have a branch on my repo soon. Should we bump this conversation to issue #9587 and let this issue close with PR #9586?

**Comment by bmw** [comment]
> Should we bump this conversation to issue https://github.com/certbot/certbot/issues/9587 and let this issue close with PR https://github.com/certbot/certbot/pull/9586?

Good idea. I'll respond to the rest of your post there.

Closing this as fixed with #9586. The changes will be available under the `latest` tag after our 2.4.0 release planned for early March.
